File: dart_cli/lib/generateGlobalKeysTable.dart

import "dart:convert";
import "dart:io";

import "ByteDataWrapper.dart";
import "structs/KeysTable.dart";

Future<void> generateGlobalKeysTableFromFiles(List<String> jsonFiles, String globalKeysTablePath) async {
  var allJsonObjects = <Object>[];
  for (var jsonFile in jsonFiles) {
    var jsonStr = await File(jsonFile).readAsString();
    var json = jsonDecode(jsonStr);
    allJsonObjects.add(json);
  }
  await generateGlobalKeysTableFromJson(allJsonObjects, globalKeysTablePath);
}

Future<void> generateGlobalKeysTableFromJson(List<Object> objects, String globalKeysTablePath) async {
  KeysTable keysTable;
  if (await File(globalKeysTablePath).exists()) {
    var keysTableBytes = await ByteDataWrapper.fromFile(globalKeysTablePath);
    keysTable = KeysTable.readBytes(keysTableBytes);
  }
  else {
    keysTable = KeysTable();
  }
  for (var object in objects)
    _visitJson(object, keysTable);
  
  var keysTableBytes = ByteDataWrapper.allocate(keysTable.size);
  keysTable.writeBytes(keysTableBytes);
  await keysTableBytes.toFile(globalKeysTablePath);
}

void _visitJson(Object? json, KeysTable keysTable) {
  if (json is Map) {
    for (var key in json.keys)
      keysTable.visitKey(key);
    for (var value in json.values)
      _visitJson(value, keysTable);
  }
  if (json is List) {
    for (var item in json)
      _visitJson(item, keysTable);
  }
}

File: python_lib/generateGlobalKeysTable.py
import os
from typing import Iterable

from FileWrapper import FileWrapper
from structs.KeysTable import KeysTable


def generateGlobalKeysTableFromJson(objects: Iterable[dict], globalKeysTablePath: str, append: bool = False):
	if append and os.path.exists(globalKeysTablePath):
		with FileWrapper.fromFile(globalKeysTablePath, "rb") as f:
			keysTable = KeysTable.readBytes(f)
	else:
		keysTable = KeysTable()

	for obj in objects:
		_visitJson(obj, keysTable)

	with FileWrapper.fromFile(globalKeysTablePath, "wb") as f:
		keysTable.writeBytes(f)

def _visitJson(json, keysTable: KeysTable):
	if isinstance(json, dict):
		for key in json.keys():
			keysTable.visitKey(key)
		for value in json.values():
			_visitJson(value, keysTable)
	if isinstance(json, list):
		for item in json:
			_visitJson(item, keysTable)

def generateGlobalKeysTableFromList(keys: Iterable[str], globalKeysTablePath: str, append: bool = False):
	if append and os.path.exists(globalKeysTablePath):
		with FileWrapper.fromFile(globalKeysTablePath, "rb") as f:
			keysTable = KeysTable.readBytes(f)
	else:
		keysTable = KeysTable()

	for key in keys:
		keysTable.visitKey(key)

	with FileWrapper.fromFile(globalKeysTablePath, "wb") as f:
		keysTable.writeBytes(f)
